clients/drcachesim/simulator/cache_simulator.cpp

```diff
@@ -118,33 +126,157 @@ cache_simulator_t::cache_simulator_t(const cache_simulator_knobs_t &knobs_) :
             success = false;
             return;
         }
+
+        cache_name = "L1_I_Cache_%u" + i;
+        all_caches[cache_name] = l1_icaches[i];
+        cache_name = "L1_D_Cache_%u" + i;
+        all_caches[cache_name] = l1_dcaches[i];
     }
 }
 
-cache_simulator_t::~cache_simulator_t()
+cache_simulator_t::cache_simulator_t(const string &config_file) :
+    simulator_t(),
+    l1_icaches(NULL),
+    l1_dcaches(NULL),
+    is_warmed_up(false),
+    is_two_level(false)
 {
-    if (llcache == NULL)
+    llcaches.clear();
+
+    std::map<string, cache_params_t> cache_params;
+    config_reader_t config_reader;
+    if (!config_reader.configure(config_file, knobs, cache_params)) {
+        ERRMSG("Usage error: Failed to read/parse configuration file %s.\n",
+               config_file.c_str());
+        success = false;
         return;
-    delete llcache->get_stats();
-    delete llcache->get_prefetcher();
-    delete llcache;
-    if (icaches == NULL)
+    }
+
+    init_knobs(knobs.num_cores, knobs.skip_refs,
+               knobs.warmup_refs, knobs.warmup_fraction,
+               knobs.sim_refs, knobs.cpu_scheduling,
+               knobs.verbose);
+
+    if (knobs.data_prefetcher != PREFETCH_POLICY_NEXTLINE &&
+        knobs.data_prefetcher != PREFETCH_POLICY_NONE) {
+        // Unknown prefetcher type.
+        success = false;
         return;
-    for (unsigned int i = 0; i < knobs.num_cores; i++) {
-        // Try to handle failure during construction.
-        if (icaches[i] == NULL)
+    }
+
+    bool warmup_enabled = ((knobs.warmup_refs > 0) ||
+                           (knobs.warmup_fraction > 0.0));
+
+    l1_icaches = new cache_t* [knobs.num_cores];
+    l1_dcaches = new cache_t* [knobs.num_cores];
+
+    // Create all the caches in the hierarchy.
+    for (auto &cache_params_it : cache_params) {
+        string cache_name = cache_params_it.first;
+        auto &cache_config = cache_params_it.second;
+
+        cache_t *cache = create_cache(cache_config.replace_policy);
+        if (cache == NULL) {
+            success = false;
             return;
-        delete icaches[i]->get_stats();
-        delete icaches[i]->get_prefetcher();
-        delete icaches[i];
-        if (dcaches[i] == NULL)
+        }
+
+        all_caches[cache_name] = cache;
+    }
+
+    // Initialize all the caches in the hierarchy and identify both
+    // the L1 caches and LLC(s).
+    for (auto &cache_it : all_caches) {
+        string cache_name = cache_it.first;
+        cache_t *cache = cache_it.second;
+
+        // Locate the cache's configuration.
+        const auto &cache_config_it = cache_params.find(cache_name);
+        if (cache_config_it == cache_params.end()) {
+            ERRMSG("Error configuring the cache: %s\n", cache_name.c_str());
```

Comment:
Could this be more descriptive: probably the next string is meant for here, "Error locating..."